sdk/core/azure-core/src/http/curl/curl.cpp

```diff
@@ -807,3 +809,37 @@ int64_t CurlSession::ResponseBufferParser::BuildHeader(
   // Parsing Headers will make sure to move one possition
   return indexOfEndOfStatusLine + 1 - buffer;
 }
+
+std::list<std::unique_ptr<CurlSession::CurlConnection>> CurlSession::c_connectionPool;
+CurlSession::CurlConnection* CurlSession::GetCurlConnection(std::string const& host)
+{
+  auto connectionIterator = c_connectionPool.begin();
+  while (connectionIterator != c_connectionPool.end())
```

Comment:
This is O(n) and can be easily optimized to O(1)

Reply:
The intention is to be able to erase connections in O(1) with a list. 
In order to make O(1) for the search, we would need a Vector right?  but then erasing connection becomes expensive.

Reply:
No, not a vector.

We only keep available connections in this list. Every time we need a new connection, we get one from this list and erase it from list. We create a new one when the list is empty.

We put the connection back into this list we finish using it.

Reply:
still need to loop the list because we need to get a connection to the same host and not just the first available connection.

We can have a connection pool per host to have O(1). But that would means a connection pool per client.  We would lost the benefit of using an existing connection to the same host amount multiple clients

Reply:
Looping over the list is very inefficient. Let's say we have 30k ongoing requests. Every time we request a new connection we have to loop over the 30k nodes, which is protected by a mutex. This is not scalable.

Reply:
> We would lost the benefit of using an existing connection to the same host amount multiple clients

I think i'm okay with this.

Reply:
> Looping over the list is very inefficient. Let's say we have 30k ongoing requests. Every time we request a new connection we have to loop over the 30k nodes, which is protected by a mutex. This is not scalable.

No. Connections are pulled out of the list and moved to the session.
In use connections are never inside the pool

The case you mentioned would require 30k sessions moving the connection back to the pool, then a new connection is requested for a different host than all of those 30k connections

This initial design does not have a mutex.

This should be enough to unblock the perf issue for now.

We might enhance this design on next releases right?

Reply:
okay, I looked through you code again. So only unused connections are in that pool. This is also what I want.

Reply:
Though I still don't think it's a good idea to leave something to next release.  Why not settle everything down in this release, it's not that complex a feature.

Reply:
Let's just get this working for now and not prematurely optimize. If we profile this and this is the perf issue, then we can optimize it. I think the perf hit here is negligible especially just before an network operation. Also, as you all noted, this is just iterating over unused connections now; not the in-use ones.

Reply:
Hi @vhvb1989 and @JeffreyRichter , storage team just talked about this connection issue internally.

C++ SDK is supposed to be used by those customers who have high-concurrency, high-performance needs. So it's quite common that they have huge amount of concurrent connections.

Let's say one customer issues 30k connections at the same time. After a while, all these 30k connections finish. After 180 seconds, storage server will close those inactive connections. In this scenario, client side must release the client ports in time, maybe in dozens seconds or in several minutes.

You **CANNOT** keep those client ports unreleased forever until the program exits. If this is the case, i'm afraid we cannot accept this fix.